game-app/game-core/src/main/java/games/strategy/triplea/delegate/power/calculator/AvailableSupports.java

```diff
@@ -164,13 +168,13 @@ int getSupportLeft(final UnitSupportAttachment support) {
    * <p>This may return the same unit multiple times in a row depending on how much support that
    * unit can give.
    */
-  private Unit getNextAvailableSupporter(final UnitSupportAttachment support) {
+  private Unit getNextAvailableSupporter(final UnitSupportAttachment support, final int i) {
     final SupportDetails details = supportUnits.get(support);
     final IntegerMap<Unit> intMap = details.supportUnits;
-    final Unit u = CollectionUtils.getAny(intMap.keySet());
+    final Unit u = CollectionUtils.getAt(intMap.keySet(), i);
```

Comment:
I'm having trouble understanding the significance of this change.
Questions:
(1) Why do we know that the size of `intMap.keySet()` is less than `i`?
(2) Why is the `i`'th element significant? In other words, why that one element and not the first or last one? I think if I spent more time with this, it might "click" and I would understand, but perhaps you can readily give insight here?

Reply:
Element stored in intMap are, to my understanding, linear. getAny will always return the first element until that element was removed. getAt returns the element at a given position, thus element 1 may still have a count but has already been used, so I want to skip to the next available element with a count.

Reply:
> (1) Why do we know that the size of `intMap.keySet()` is less than `i`?

The size of `intMap.keySet()` should be either greater than or equal to `i` and is determined prior to calling getNextAvailableSupporter by `numSupportAvailableToApply = getSupportersAvailable(rule)`. And because getNextAvailableSupporter can decrease the size, it is again rechecked with `int count = Math.min(i, getSupportersAvailable(rule))`. 



> (2) Why is the `i`'th element significant? In other words, why that one element and not the first or last one? I think if I spent more time with this, it might "click" and I would understand, but perhaps you can readily give insight here?

This is the **problem**, IMHO, that we are having right now. `CollectionUtils.getAny(intMap.keySet())` only returns the first element for `u`, and does not consider if that element for might have already been use prior in the loop.

As of now giveSupportToUnit uses the sun total of **number** and is restricted by **bonusType/count**, and does not care how many units are available. Such that a unit can give as much support as limited by **bonusType/count** to a given unit on a 1 to x basis as determined by **bonusType/count**.
This changes giveSupportToUnit to use the number of units available to give support then restricts that by **bonusType/count** and then tries to choose different supporting units. In assents, a unit can give support to another unit on a 1 to 1 basis.

Hope this makes things clearer.

Cheers...

Reply:
Just revisiting this, but another option that doesn't have the issue with getAt() discussed in the other thread (i.e. order not guaranteed) is to actually keep the state in an Iterator object. That way, you can just advance through it via next() etc.